cleverhans/feat_advs.py

```diff
@@ -0,0 +1,151 @@
+"""
+This module inmplements a fast implementation of Feature Adversaries, an attack
+against a target internal representation of a model.
+Feature adversaries were originally introduced in (Sabour et al. 2016),
+where the optimization was done using LBFGS.
+Paper link: https://arxiv.org/abs/1511.05122
+"""
+import numpy as np
+import tensorflow as tf
+
+from cleverhans.attacks import Attack
+from cleverhans.model import Model
+
+
+class FastFeatureAdversaries(Attack):
+
+    """
+    This is similar to Basic Iterative Method (Kurakin et al. 2016) but
+    applied to the internal representations.
+    """
+
+    def __init__(self, model, back='tf', sess=None):
+        """
+        Create a FastFeatureAdversaries instance.
+        """
+        super(FastFeatureAdversaries, self).__init__(model, back, sess)
+        self.feedable_kwargs = {'eps': np.float32,
+                                'eps_iter': np.float32,
+                                'clip_min': np.float32,
+                                'clip_max': np.float32,
+                                'layer': str}
+        self.structural_kwargs = ['ord', 'nb_iter']
+
+        assert isinstance(self.model, Model)
+
+    def parse_params(self, layer=None, eps=0.3, eps_iter=0.05, nb_iter=10,
+                     ord=np.inf, clip_min=None, clip_max=None, **kwargs):
+        """
+        Take in a dictionary of parameters and applies attack-specific checks
+        before saving them as attributes.
+
+        Attack-specific parameters:
+        :param layer: (required str) name of the layer to target.
+        :param eps: (required float) maximum distortion of adversarial example
+                    compared to original input
+        :param eps_iter: (required float) step size for each attack iteration
+        :param nb_iter: (required int) Number of attack iterations.
+        :param ord: (optional) Order of the norm (mimics Numpy).
+                    Possible values: np.inf, 1 or 2.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+
+        # Save attack-specific parameters
+        self.layer = layer
+        self.eps = eps
+        self.eps_iter = eps_iter
+        self.nb_iter = nb_iter
+        self.ord = ord
+        self.clip_min = clip_min
+        self.clip_max = clip_max
+
+        # Check if order of the norm is acceptable given current implementation
+        if self.ord not in [np.inf, 1, 2]:
+            raise ValueError("Norm order must be either np.inf, 1, or 2.")
+        if self.back == 'th':
```

Comment:
You can remove this given that we dropped Theano support.

Reply:
Removed in 30b4472.